### ISSUE TITLE
Support for Golang

### ISSUE THREAD

**Issue by traxair** [swagger-api/swagger-codegen#325]
Is it in the backlog?

**Comment by mattolenik** [comment]
Is any of this work yet available (even if untested)? I'm on a project where this would be useful, and I'm contemplating writing Golang support for Swagger.

**Comment by fehguy** [comment]
I suggest joining forces with @casualjim who is very far along with his [project](https://github.com/casualjim/go-swagger).

**Comment by mattolenik** [comment]
I'd rather just implement a generator for the existing codegen, rather than write something from the ground up in Go.

Unfortunately, I've run into a problem. I'm not sure if I the codegen model supports being able to generate idiomatic Go code. What I want is to have a Go package for each version of each API, and in that package will be the code for the models and the operations. I don't want separate modules for operations and models. Models and operations need to go into the same directory, and that directory needs to be based on the operation name. For example, if I have some operation "foo" then you'd expect to see all the code under "src/my/api/foo/v1", and you'd expect each .go source file in this directory to say "package foo" at the top.

Does this make sense? Is there any way I can do this with the current codegen model? This is a good example of the structure I'm trying to mimmic: https://github.com/google/google-api-go-client

**Comment by fehguy** [comment]
Fork swagger-codegen, send me the URL and I'll send a pull request to help with the structure.

**Comment by casualjim** [comment]
There are a bunch of reasons why I didn't do it in this codegen package. And mostly it's very challenging for me to get full support for the swager spec in the java version of swagger codegen.
I started out trying to customize this codegen instead and only after a bunch of fruitless attempts I went on with creating the entire generator stuff in go-swagger.
Go types are not as clear cut as it seems, when doing codegen. Then making default values is again not entirely straight forward in java. Struct literals etc all that stuff you kind of get for free when you use go.

At first i tried to keep it so I could at least reuse mustache templates but none of the go mustache packages I could find allows for self-referencing partial mustache templates. And it's something I needed because there are many validations etc that need to be respected.  Again I tried many other ways to make mustache work, before moving on to the go template system.

Now if you can make it work all the better but those are the reasons I didn't/couldn't.
I also want to provide a reverse codegen that generates a swagger spec based on your go files. 

And then I thought it made sense to have one command that helps you in dealing with all kinds of things swagger.

**Comment by mattolenik** [comment]
@casualjim, does your project currently do client generation? That's what I need at the moment.

@fehguy, thanks, but this is part of a project at work, so that's not really an option.

**Comment by casualjim** [comment]
not at present. I have a couple of things left to fix with the server. if nobody else contributes to it then I should be able get to it in a week or so.

**Comment by alfonsodev** [comment]
I could help with the golang client, I forked and create a golang_client_support branching from develop_2.0  https://github.com/alfonsodev/swagger-codegen/tree/golang_client_support  
@fehguy can you send me a pull request with the structure ? If I have to only work with templates I'm pretty sure I can do it.  
thanks!

**Comment by wing328** [comment]
@alfonsodev thanks for offering help for the Golang client.

To start with, I would suggest you look at the work done by @casualjim as a reference. His latest PR (not merged) for another project (swagger-go) should be a good starting point - https://github.com/go-swagger/go-swagger/pull/6.

Let me know if you need any help or have any questions..

**Comment by casualjim** [comment]
The client generation has been merged to master now.

**Comment by wing328** [comment]
:+1: Thanks @casualjim

**Comment by wing328** [comment]
FYI. I submitted #1747 to add a Go API client generator.

**Comment by wing328** [comment]
#1747 (A beta Go API client) has been merged into master. Please take a look and let us know if you've any feedback.

The Go API client is not perfect and if you would like to contribute to make it better, please refer to the following list of tasks that are pending community contribution.

https://github.com/swagger-api/swagger-codegen/issues?q=is%3Aopen+is%3Aissue+label%3A%22Client%3A+Go%22
